vortex-expr/src/merge.rs

```diff
@@ -0,0 +1,323 @@
+use std::any::Any;
+use std::fmt::Display;
+use std::hash::Hash;
+use std::sync::Arc;
+
+use itertools::Itertools as _;
+use vortex_array::array::StructArray;
+use vortex_array::validity::Validity;
+use vortex_array::{ArrayDType, ArrayData, IntoArrayData};
+use vortex_dtype::FieldNames;
+use vortex_error::{vortex_bail, VortexExpect as _, VortexResult};
+
+use crate::{ExprRef, VortexExpr};
+
+/// Merge zero or more expressions that ALL return structs.
+///
+/// If any field names are duplicated, the field from later expressions wins.
+///
+/// NOTE: Fields are not recursively merged, i.e. the later field REPLACES the earlier field.
+/// This makes struct fields behaviour consistent with other dtypes.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Merge {
+    values: Vec<ExprRef>,
+}
+
+impl Merge {
+    pub fn new_expr(values: Vec<ExprRef>) -> Arc<Self> {
+        Arc::new(Merge { values })
+    }
+}
+
+pub fn merge(elements: impl IntoIterator<Item = impl Into<ExprRef>>) -> ExprRef {
+    let values = elements.into_iter().map(|value| value.into()).collect_vec();
+    Merge::new_expr(values)
+}
+
+impl Display for Merge {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("{")?;
+        self.values
+            .iter()
+            .format_with(", ", |expr, f| f(expr))
+            .fmt(f)?;
+        f.write_str("}")
+    }
+}
+
+impl VortexExpr for Merge {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn unchecked_evaluate(&self, batch: &ArrayData) -> VortexResult<ArrayData> {
+        let len = batch.len();
+        let value_arrays = self
+            .values
+            .iter()
+            .map(|value_expr| value_expr.evaluate(batch))
+            .process_results(|it| it.collect::<Vec<_>>())?;
+
+        // Collect fields in order of appearance. Later fields overwrite earlier fields.
+        let mut field_names = Vec::new();
+        let mut arrays = Vec::new();
+
+        for value_array in value_arrays.iter() {
+            // Expect non-nullable struct input, for now.r
```

Comment:
Is this for now, could you explain, this in the rust doc and how to expect struct with validity to behave, once implemented.

Reply:
Updated to use todo instead of error. That will make it clear that's temporary limitation.